src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Collect.cs

```diff
@@ -0,0 +1,472 @@
+            var firstNameNode = args[0]?.AsFirstName();
+            identifierNode = firstNameNode;
+
+            if (firstNameNode == null)
+            {
+                return null;
+            }
+
+            var identifiers = new List<Identifier>
+            {
+                firstNameNode.Ident
+            };
+            return identifiers;
+        }
+
+        public override bool IsAsyncInvocation(PowerFx.Syntax.CallNode callNode, TexlBinding binding)
+        {
+            Contracts.AssertValue(callNode);
+            Contracts.AssertValue(binding);
+
+            return Arg0RequiresAsync(callNode, binding);
+        }
+
+        public static DType GetCollectedTypeForGivenArgType(Features features, DType argType, TexlNode arg, ref Dictionary<TexlNode, DType> nodeToCoercedTypeMap)
+        {
+            var singleColumnRecordType = GetCollectedTypeForGivenArgType(features, argType);
+
+            if (!argType.IsPrimitive)
+            {
+                return argType;
+            }
+
+            //CollectionUtils.Add(ref nodeToCoercedTypeMap, arg, singleColumnRecordType);
+
+            return singleColumnRecordType;
+        }
+
+        public static DType GetCollectedTypeForGivenArgType(Features features, DType argType)
+        {
+            Contracts.Assert(argType.IsValid);
+
+            if (!argType.IsPrimitive)
+            {
+                return argType;
+            }
+
+            // Passed a scalar; make a record out of it, using a name that depends on the type.
+            var fieldName = Contracts.VerifyValue(CreateInvariantFieldName(features, argType.Kind));
+            return DType.CreateRecord(new TypedName[] { new TypedName(argType, new DName(fieldName)) });
+        }
+
+        protected static string CreateInvariantFieldName(PowerFx.Features features, DKind dKind)
+        {
+            Contracts.Assert(dKind >= DKind._Min && dKind < DKind._Lim);
+
+            return MutationUtils.GetScalarSingleColumnNameForType(features, dKind);
+        }
+    }
+
+    // Collect(collection:*[...], item1, ...)
+    internal class CollectScalarFunction : CollectFunction
+    {
+        public override bool IsScalar => true;
+
+        public override bool SupportsParamCoercion => false;
+
+        // This method returns the name of the record to be used when a scalar is passed to Collect and converted.
+        // It is critical that these are *always* the invariant names.
+        public static string GetInvariantNameForRecord(PowerFx.Features features, DKind dKind)
+        {
+            return CreateInvariantFieldName(features, dKind);
+        }
+
+        public override DType GetCollectedType(Features features, DType argType)
+        {
+            return GetCollectedTypeForGivenArgType(features, argType);
+        }
+
+        internal override IntermediateNode CreateIRCallNode(PowerFx.Syntax.CallNode node, IRTranslator.IRTranslatorContext context, List<IntermediateNode> args, ScopeSymbol scope)
```

Comment:
>CreateIRCallNode [](http://example.com/codeflow?start=43&length=16)

This is interesting. 

@jas-valgotar , fyi

Reply:
What is this doing? can you also add comment?
Is it Converting
Collect(t, 1, 2,) -> Collect(t, {Value: 1}, {Value: 2}) ?

Reply:
Also is there a test for it?

Reply:
> What is this doing? can you also add comment? Is it Converting Collect(t, 1, 2,) -> Collect(t, {Value: 1}, {Value: 2}) ?

Correct. PA handles scalar values at runtime. We are transforming scalars into `{Value:n}` in IR.